### PR TITLE
Add functional collective all_to_all_single and support it in Inductor

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110195

Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by ezyang** [comment]
cc @yf225

**Comment by ezyang** [comment]
Hmm I think I didn't do the merge update completely correctly

File: test/distributed/test_inductor_collectives.py [hunk 1]
```diff
@@ -261,6 +267,170 @@ def compile(func, example_inputs):
...
+                torch.ones(int(row), 5, device="cuda") * (self.rank + 1),
+                input_split_sizes_tensor,
+                output_split_sizes_tensor,
+            )
+            trs = self.get_world_trs()
+
+            compiled_fn = torch.compile(example, fullgraph=True, dynamic=True)
+            code = run_and_get_triton_code(compiled_fn, *inputs, **trs)
+            FileCheck() \
+                .check("fun_col_impl._all_to_all_single") \
```
Comment (yf225) [inline test/distributed/test_inductor_collectives.py:312]: should we assert that generated code has the unbacked symints at the right places?

Comment (ezyang) [reply test/distributed/test_inductor_collectives.py:312]: it was a pain manually updating the FileChecks when I wobbled the code lol. Can we come up with a less brittle way to test this

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -4896,6 +4930,7 @@ def get_strides_of_lstm_output(output_shape, batch_first):
                 zip(output_sizes, output_strides)
             )
         ]
+        packed.outputs = output_ir
```
Comment (ezyang) [inline torch/_inductor/ir.py:4933]: I guess I don't need this anymore

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -509,6 +541,17 @@ def mk_out_tensor(input):
 
     return [mk_out_tensor(t) for t in inputs]
 
+# TODO: think, this isn't actually dynamic?!
```
Comment (ezyang) [inline torch/distributed/_functional_collectives.py:544]: I was surprised here

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -2329,6 +2329,8 @@ def _init(
         # Maps symbolic ints to their original concrete values
         # Currently populated from tensors
         self.var_to_val: Dict[sympy.Symbol, sympy.Integer] = {}
+        # Maps symbolic ints to their size hints. Only used for unbacked symints.
+        self.var_to_size_hint: Dict[sympy.Symbol, sympy.Integer] = {}
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2333]: still not convinced by this

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -2486,6 +2524,23 @@ def get_unbacked_symbol_uses(self):
         """
         return set()
 
+    def codegen_unbacked_symbol_defs(self, wrapper):
+        symbols_to_define = self.get_unbacked_symbol_defs()
```
Comment (Skylion007) [inline torch/_inductor/ir.py:2528]: Hmm, if it wasn't for the logging here, I would advise rewriting this as set intersection / difference ops.

**Review by Skylion007 — APPROVED** [review]

File: test/distributed/test_inductor_collectives.py (hunk 1 above)
Comment (yf225) [reply test/distributed/test_inductor_collectives.py:312]: perhaps a regex matching the prefix `i` but not the actual variable number? I can add it after this PR

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -6213,3 +6256,106 @@ def maybe_free_unbacked_symbols(s):
...
+            r |= free_unbacked_symbols(self.output_split_sizes)
+        if self.input_split_sizes is not None:
+            r |= free_unbacked_symbols(self.input_split_sizes)
+        return r
+
+    def should_emit_register_tensor_work(self):
+        return False
+
+    def should_emit_find_or_create_pg(self):
+        return False
```
Comment (ezyang) [inline torch/_inductor/ir.py:6297]: @yf225 can you explain why you needed to disable register tensor work / find-create pg? This seems surprising to me

**Comment by ezyang** [comment]
@yf225 because the kernel is no longer "dynamic", it can be implemented exactly symmetrically to all the other collectives. I've updated the implementation accordingly.

**Pull request closed without merging**